corehq/apps/reports/tests/test_tableau_api.py

```diff
@@ -1,34 +1,328 @@
+                            {
+                                "domain": {
+                                    "name": "local"
+                                },
+                                "id": "%s",
+                                "name": "%s"
+                            }
+                        ]
+                    }
+                }
+                """ % (self.group_ids[0], self.group_names[0], self.group_ids[1], self.group_names[1]))
+
+    def create_user_response(self, username, role):
+        return self._create_response("""
+            {
+                "user": {
+                    "id": "gh23jk",
+                    "name": "%s",
+                    "siteRole": "%s",
+                    "authSetting": "None"
+                }
+            }
+        """ % (username, role))
+
+    def update_user_response(self, **kwargs):
+        return self._create_response('')
+
+    def delete_user_response(self, **kwargs):
+        return self._create_response('')
+
+    def sign_out_response(self, **kwargs):
+        return self._create_response('')
+
+    def failure_response(self):
+        response_text = """
+            {
+                "error": {
+                    "summary": "Invalid Authentication Credentials",
+                    "detail": "Your authentication credentials are invalid.",
+                    "code": "401002"
+                }
+            }
+        """
+        mock_response = requests.Response()
+        mock_response.status_code = 400
+        mock_response._content = response_text.encode('utf-8')
+        return mock_response
+
+
+class TestTableauAPI(TestCase):
 
     def setUp(self):
         self.domain = 'test-domain-name'
         self.test_server = TableauServer(
             domain=self.domain,
             server_type='server',
-            server_name='server name',
+            server_name='test_server',
             validate_hostname='host name',
             target_site='target site'
         )
-        self.test_connected_app = TableauConnectedApp(
+        self.test_server.save()
+        self.connected_app = TableauConnectedApp(
             app_client_id='asdf1234',
             secret_id='zxcv5678',
             server=self.test_server
         )
-        super(TestTableauConnectedApp, self).setUp()
+        self.connected_app.save()
+        self.tableau_instance = FakeTableauInstance()
+        super(TestTableauAPI, self).setUp()
+
+    def test_connected_app_encryption(self):
+        self.connected_app.plaintext_secret_value = 'qwer1234'
+        self.connected_app.save()
+
+        self.assertNotEqual(self.connected_app.encrypted_secret_value, 'qwer1234')
+        self.assertEqual(self.connected_app.plaintext_secret_value, 'qwer1234')
+        self.assertEqual(len(self.connected_app.encrypted_secret_value), 24)
+
+    def _assert_subset(self, d1, d2):
+        self.assertTrue(set(d1.items()).issubset(set(d2.items())))
+
+    def test_tableau_API_session(self):
```

Comment:
could this be split into multiple tests (one for each 'method')?

Reply:
I would say that all the tests should include "getting the API version and signing in" but you should still split out the other parts which are independent of each-other.

So you could have individual tests for `query_groups`, `get_users_in_group`, `create_group`, `add_user_to_group`, `get_groups_for_user_id`, `create_user`, `update_user`, `delete_user`.